coral/src/app/features/topics/acl-request/TopicAclRequest.tsx

```diff
@@ -52,34 +59,41 @@ const TopicAclRequest = () => {
     },
   });
 
-  const { data: topicNames } = useQuery<TopicNames, Error>(
-    ["topic-names", topicType],
-    {
-      queryFn: () => getTopicNames({ onlyMyTeamTopics: false }),
-      keepPreviousData: true,
-      onSuccess: (data) => {
-        if (data?.includes(topicName)) {
-          return;
-        }
-        // Navigate back to Topics when topicName does not exist in the topics list
-        navigate("/topics");
-      },
-      enabled: topicName !== "",
-    }
-  );
+  const { data: environments, isLoading: environmentsIsLoading } = useQuery<
+    Environment[],
+    Error
+  >(["topic-environments"], {
+    queryFn: getEnvironments,
+  });
 
-  const { data: environments } = useQuery<Environment[], Error>(
-    ["topic-environments"],
-    {
-      queryFn: getEnvironments,
-    }
-  );
+  const topicNamesQueries =
```

Comment:
Not extremely hapy with this. I would be grateful for feedbacl on to how to do this in a more legible way. Ideas:

- somehow compute all those queries as part of the `onSuccess` of the `getEnvironments` query
- break the `getEnvironment` and `getTopicNames` queries in a single custom hook query that would return the `validEnvironments` and `scopedTopicNames` instead of having all that business logic inline in this component

Reply:
To summarise:

Before we can decide what topicNames to show in the dropdown, an environment has to be chosen, right? 🤔 

So... user choses env -> option of topic name change to the ones available? 

If that's the case and it's not more complex - would it be an idea to only fetch the `topicNames` after an environment is selected? And ideally also only when "Topic pattern type" is selected? Will lead to a moment of waiting for the user to get the topic select filled with options, but it would make the logic the frontend has to handle less complex. Also, it would reduce the amount of api calls we have to do to a minimum. 

Alternative I also would look into moving this functionality in a custom hook (or at least out of this component). It's already quite complex and not super easy to understand what's happening :D 

(ideally, in long run, this is something the backend could provide us a more useful data structure for)

Reply:
> If that's the case and it's not more complex - would it be an idea to only fetch the topicNames after an environment is selected? And ideally also only when "Topic pattern type" is selected? Will lead to a moment of waiting for the user to get the topic select filled with options, but it would make the logic the frontend has to handle less complex. Also, it would reduce the amount of api calls we have to do to a minimum.

"Fetch as you go" was my first approach but it ends up actually being less pleasant to work with, and a less pleasant UX, which is why I chose this frontloading approach. Most notably, testing gets *super flaky* when that kind of interaction triggers an API request, which results in a re-render of another part of the UI.

But I agree that this should be a hook and not inline. And also that the fact we need to do that kind of computation is a sign that the backend should provide a little more ^^

Reply:
TBH I almost want to do it for `getClusterInfo` as well, it's super jarring to have the UI react after 500ms/1000ms to data returned from the API and have field appear / disappear. We could batch it the same way as a `useIsAivenCluster`  query hook.

Here this is throttled to fast 4G, and you can see how weird it feels to see the `Transactional ID` and other UI state take a little bit to update, while the `Select` input is immediately updated with the relevant topic names.

https://user-images.githubusercontent.com/20607294/216004613-6338fe11-d3c3-4689-8d41-fb89618d8cca.mov

Reply:
Seems way neater this way :o [3cdce9e](https://github.com/aiven/klaw/pull/518/commits/3cdce9e4dc86e6f77d93d77d5c8d78cff373c1c7)

Reply:
> TBH I almost want to do it for getClusterInfo as well

I did it for it as well and it's *so much better*. There is no lag in the form rendering now, and I expect all the flakyness is gone from the tests.

Reply:
Comparison of the difference the "pre fetching everything" appraoch makes:

### Before

We fetch as we go, so there will always be latency on the first fetch. Results are cached for subsequent calls, so it's snappy, but the initial experience is pretty bad.

https://user-images.githubusercontent.com/20607294/216020607-963435ac-4409-4ab5-bdf1-c6ea5b5008ef.mov

### After

It includes work around `getClusterInfo` not currently on this branch, because it probably be on a separate PR. Snappy from the get go, with probably a fractionally longer loading state initially.

https://user-images.githubusercontent.com/20607294/216021296-bd2ed574-0b49-4eb9-aaf3-1917734277f0.mov

Reply:
Nice ✨  Agree, that is a way better UX!